### PR TITLE
Ica bench

### PR DESCRIPTION
**Pull request by oojo12**
### Additions
Adds benchmarking code for linfa-ica it is a small refactor of the provided example. Will add benchmark stats if accepted.

### CODE REVIEW

File: algorithms/linfa-ica/benches/fast_ica.rs
Comment (YuhanLiin) [inline algorithms/linfa-ica/benches/fast_ica.rs:12]: Also run separate benchmarks with `GFunc::Exp` and `GFunc::Cube`.

Also set `random_seed` for reproducible results.

File: algorithms/linfa-ica/benches/fast_ica.rs
Comment (YuhanLiin) [inline algorithms/linfa-ica/benches/fast_ica.rs:9]: Don't need to specify return type

File: algorithms/linfa-ica/benches/fast_ica.rs
Comment (YuhanLiin) [inline algorithms/linfa-ica/benches/fast_ica.rs:51]: How long does 100_000 samples take? It always takes too long on my machine, but maybe ICA is a lot faster than other algorithms.

File: algorithms/linfa-ica/benches/fast_ica.rs
Comment (oojo12) [reply algorithms/linfa-ica/benches/fast_ica.rs:12]: lines 38-40 have the random_seed set to add reproducible noise. The original signals are created from 

`Array::linspace(0., 8., nsamples)` which should guarantee reproducibility. 

I'm adding the additional functions now though.

File: algorithms/linfa-ica/benches/fast_ica.rs
Comment (oojo12) [reply algorithms/linfa-ica/benches/fast_ica.rs:51]:
```
fast_ica_bench/fast-ica-{}/100000
                        time:   [15.202 ms 15.575 ms 15.967 ms]
Found 5 outliers among 100 measurements (5.00%)
  5 (5.00%) high mild

```

File: algorithms/linfa-ica/benches/fast_ica.rs
Comment (YuhanLiin) [reply algorithms/linfa-ica/benches/fast_ica.rs:12]: By `random_seed` I meant calling `FastIca::params().random_seed()` to set a reproducible seed for the ICA algorithm itself.

File: algorithms/linfa-ica/benches/fast_ica.rs
Comment (YuhanLiin) [reply algorithms/linfa-ica/benches/fast_ica.rs:51]: Nice

File: algorithms/linfa-ica/benches/fast_ica.rs
Comment (oojo12) [reply algorithms/linfa-ica/benches/fast_ica.rs:28]: oh gotcha will do now.

**Comment by oojo12** [comment]
Example output

```
test result: ok. 0 passed; 0 failed; 6 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running benches\fast_ica.rs (target\release\deps\fast_ica-239ba21dab7bb851.exe)
Fast ICA/GFunc_LogCosH/1000
                        time:   [74.308 µs 74.549 µs 74.808 µs]
Found 10 outliers among 100 measurements (10.00%)
  9 (9.00%) high mild
  1 (1.00%) high severe
Fast ICA/GFunc_LogCosH/10000
                        time:   [745.54 µs 747.69 µs 749.94 µs]
Found 10 outliers among 100 measurements (10.00%)
  1 (1.00%) low mild
  4 (4.00%) high mild
  5 (5.00%) high severe
Fast ICA/GFunc_LogCosH/100000
                        time:   [10.645 ms 10.671 ms 10.700 ms]
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe
```

**Comment by oojo12** [comment]
not sure if the test is flaky but when I run `cargo clippy --all-targets -- -D warnings` I don't get any errors as reported in the failing codequality Lint

Screenshot 2022-11-04 190052

File: algorithms/linfa-ica/benches/fast_ica.rs
Comment (YuhanLiin) [inline algorithms/linfa-ica/benches/fast_ica.rs:33]: These 3 can be refactored into a single function `perform_ica(size: usize, gfunc: GFunc)`

File: algorithms/linfa-ica/benches/fast_ica.rs
Comment (YuhanLiin) [inline algorithms/linfa-ica/benches/fast_ica.rs:95]: These 3 can be refactored into one single benchmark function. Just add an extra nested loop that goes through each gfunc, like this: `for (gfunc, name) in [(GFunc::Cube, "GFunc_Cube"), ...]`

**Comment by YuhanLiin** [comment]
You're not seeing the error because the version of Rust in CI is newer than your local version. Just fix the error so the CI passes.

File: algorithms/linfa-ica/benches/fast_ica.rs
Comment (oojo12) [reply algorithms/linfa-ica/benches/fast_ica.rs]: addressed.

**Comment by oojo12** [comment]
Gotcha it appears to be a new problem on

> You're not seeing the error because the version of Rust in CI is newer than your local version. Just fix the error so the CI passes.

Gotcha it appears to be a new problem on Master too. Assuming it happened with the Rust release Nov 3rd. I addressed it and reran the test locally with no issues with the latest Rust.

**Review by YuhanLiin — APPROVED** [review]

**Comment by YuhanLiin** [comment]
Can you do the benchmark comparison in #228 ?

**Comment by oojo12** [comment]
Sure thing

**Comment by oojo12** [comment]
I don't have permission to merge per "Only those with [write access](https://docs.github.com/articles/what-are-the-different-access-permissions) to this repository can merge pull requests." However, below are the necessary things that can be dumped in that issue once you merge. Also, the report is attached as a zip file.

### Context

1. Laptop mid charge level with wall charger power plugged in
2. Laptop not in use during test
3. Laptop not over heating

### Run: cargo bench -p linfa-ica 
```
     Running benches\fast_ica.rs (target\release\deps\fast_ica-d0c3b5752e019610.exe)
Fast ICA/GFunc_Cube/1000
                        time:   [63.292 µs 63.701 µs 64.270 µs]
Found 24 outliers among 100 measurements (24.00%)
  22 (22.00%) high mild
  2 (2.00%) high severe
Fast ICA/GFunc_Cube/10000
                        time:   [594.96 µs 599.06 µs 603.51 µs]
Found 6 outliers among 100 measurements (6.00%)
  5 (5.00%) high mild
  1 (1.00%) high severe
Fast ICA/GFunc_Cube/100000
                        time:   [10.097 ms 10.163 ms 10.236 ms]
Found 8 outliers among 100 measurements (8.00%)
  4 (4.00%) high mild
  4 (4.00%) high severe

Fast ICA/GFunc_Logcosh/1000
                        time:   [80.528 µs 80.737 µs 80.965 µs]
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) low severe
  5 (5.00%) low mild
  1 (1.00%) high mild
  1 (1.00%) high severe
Fast ICA/GFunc_Logcosh/10000
                        time:   [790.09 µs 792.99 µs 796.68 µs]
Found 13 outliers among 100 measurements (13.00%)
  1 (1.00%) low mild
  7 (7.00%) high mild
  5 (5.00%) high severe
Fast ICA/GFunc_Logcosh/100000
                        time:   [11.175 ms 11.241 ms 11.311 ms]
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

Fast ICA/Exp/1000       time:   [95.226 µs 95.656 µs 96.168 µs]
Found 11 outliers among 100 measurements (11.00%)
  6 (6.00%) high mild
  5 (5.00%) high severe
Fast ICA/Exp/10000      time:   [930.76 µs 935.38 µs 941.12 µs]
Found 12 outliers among 100 measurements (12.00%)
  1 (1.00%) low severe
  1 (1.00%) low mild
  7 (7.00%) high mild
  3 (3.00%) high severe
Fast ICA/Exp/100000     time:   [18.221 ms 18.381 ms 18.559 ms]
Found 9 outliers among 100 measurements (9.00%)
  2 (2.00%) high mild
  7 (7.00%) high severe
```

### Run cargo bench -p linfa-ica -F intel-mkl-static
```
     Running benches\fast_ica.rs (target\release\deps\fast_ica-2e03c80ee39d2cab.exe)
Fast ICA/GFunc_Cube/1000
                        time:   [51.978 µs 52.182 µs 52.431 µs]
                        change: [-35.893% -31.248% -26.694%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  9 (9.00%) high mild
  4 (4.00%) high severe
Fast ICA/GFunc_Cube/10000
                        time:   [466.80 µs 469.63 µs 473.14 µs]
                        change: [-21.665% -20.949% -20.232%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 9 outliers among 100 measurements (9.00%)
  1 (1.00%) low mild
  6 (6.00%) high mild
  2 (2.00%) high severe
Fast ICA/GFunc_Cube/100000
                        time:   [9.2229 ms 9.3005 ms 9.3897 ms]
                        change: [-9.4661% -8.4831% -7.3959%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  5 (5.00%) high mild
  6 (6.00%) high severe

Fast ICA/GFunc_Logcosh/1000
                        time:   [69.136 µs 69.490 µs 69.884 µs]
                        change: [-13.958% -13.385% -12.795%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  3 (3.00%) low mild
  4 (4.00%) high mild
  1 (1.00%) high severe
Fast ICA/GFunc_Logcosh/10000
                        time:   [672.37 µs 675.86 µs 680.20 µs]
                        change: [-15.858% -14.933% -13.922%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 20 outliers among 100 measurements (20.00%)
  2 (2.00%) low mild
  4 (4.00%) high mild
  14 (14.00%) high severe
Fast ICA/GFunc_Logcosh/100000
                        time:   [10.314 ms 10.412 ms 10.527 ms]
                        change: [-8.3714% -7.3728% -6.1955%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  3 (3.00%) high mild
  5 (5.00%) high severe

Fast ICA/Exp/1000       time:   [76.481 µs 76.798 µs 77.180 µs]
                        change: [-20.431% -19.929% -19.446%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  4 (4.00%) high mild
  9 (9.00%) high severe
Fast ICA/Exp/10000      time:   [708.61 µs 714.43 µs 721.13 µs]
                        change: [-23.322% -22.474% -21.608%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) low mild
  4 (4.00%) high mild
Fast ICA/Exp/100000     time:   [16.429 ms 16.604 ms 16.795 ms]
                        change: [-10.997% -9.6641% -8.3325%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 9 outliers among 100 measurements (9.00%)
  6 (6.00%) high mild
  3 (3.00%) high severe
```
[criterion.zip](https://github.com/rust-ml/linfa/files/9944340/criterion.zip)

**Pull request merged**
